doc: Added documentation on the external SUPL client library (#2009)
NordicPlayground/fw-nrfconnect-nrf: include/libsupl.rst

.. _libsupl:

> LibSUPL

Reviewer: Naming wise, the library part that you download as a separate component should be refereed to as the external SUPL client library, that aligns with the location in the three (ext/lib/bin/supl). And the other part as SUPL integration library (lib/supl).

> LibSUPL

Author: Shall we change the structure and merge this file with supl_os_client documentation according to the following manner?(With all references to libSUPL changing to External SUPL client library)

- Intro : 
The Secure User-Plane Location (SUPL) client can be used to retrieve A-GPS_ data from a SUPL server using the SUPL protocol.
The SUPL client module consists of an external SUPL client library and the SUPL integration library which integrates it into the nRF Connect SDK.

- contents of supl_os_client (till API) 
- overview and the rest of the contents from this file (currently named libSUPL)
- API from supl_os_client

#######

LibSUPL is an external SUPL (Secure User Plane Location) client library, which is used to retrieve `A-GPS`_ data from a SUPL server using the SUPL protocol.

.. |lib_note| replace:: Therefore, you must download and install it separately.

.. include:: supl_os_client.rst
   :start-after: supl_details_start
   :end-before: supl_details_end

See :ref:`download_supl` for information on downloading the external SUPL client library.

Overview
********

> SUPL is an Enabler, which enables the transfer of assistance and network based positioning data of a SET (SUPL Enabled Terminal) over an IP User Plane bearer.

Reviewer 2: ```suggestion
SUPL enables the transfer of assistance and network based positioning data of a SUPL Enabled Terminal (SET) over an IP User Plane bearer.
```

Reviewer 2: I'm trying to understand this - what would a SET be in our case? Would that be the DK or Thingy? Then we should add that information:

```suggestion
SUPL enables the transfer of assistance and network based positioning data of a SUPL Enabled Terminal (SET), such as an nRF9160 DK or a Thingy:91, over an IP User Plane bearer.
```

Reviewer 2: @evenl Is the SET the DK or Thingy?

See `SUPL context model`_ for more information on the SUPL architecture.

> SUPL is also designed to support future positioning technologies.

Reviewer 2: ```suggestion
SUPL is designed to support future positioning technologies.
```

> SUPL uses A-GPS for obtaining GPS data, overcoming the limitations of standalone GPS.

Reviewer: Should we list some limitations?

Author: The limitations are discussed in the subsequent paragraph. I shall modify that as follows to be more specific. 
A standalone GPS uses the orbital data (like `Ephermeris`_ and the `Almanac`_) of the satellite to calculate the current position of a device. It has the following limitations:

- Typically, the data download from the satellite takes a long time and the first position is calculated in 30-40 seconds.

- Any loss of signal requires the whole process to be restarted.

> A standalone GPS uses the orbital data (like `Ephermeridus`_ and the `Almanac`_) of the satellite to calculate the current position of a device.

Reviewer 2: Ephemeris?

Author: changed

Typically, the data download from the satellite takes a long time and the first position is calculated in 30-40 seconds.
Any loss of signal requires the whole process to be restarted.

An A-GPS server is a cached server that stores the orbital data.
This information can be downloaded at a fast rate since an A-GPS capable device supports mobile radio-network bearers with high data rate such as GSM or LTE.

> The SUPL version supported by this library is 2.0.4.

Reviewer 2: Move this line after line 20 - let's first give all information about SUPL and then move on to A-GPS.

Author: Moved.

> LibSUPL payload

Reviewer 2: Is this specific to this implementation of the library, or is it how a payload always looks when using LPP?

***************

`LPP (LTE Positioning Protocol)`_ is the supported SUPL payload format.

> LPP is a point-to-point protocol used between a location server (for example, SLP) and a target device (for example, SET) to position the device using position-based information obtained from reference sources.

Reviewer 2: What is SLP?

> A SET or SUPL Enabled Terminal is a logical part of a device that supports communication with a SUPL network.

Reviewer 2: SET has already been introduced as abbreviation, so no need to give both versions here.

> LibSUPL session and transaction type

Reviewer 2: Is this specific to this implementation or is it general for SUPL?

************************************

A typical libSUPL session is SET-based, i.e. the position is calculated based on SET.
See `SET-initiated SUPL collaboration`_ for more information.

Each LPP session consists of one or more LPP transactions.
Each LPP transaction corresponds to a single operation such as capability exchange, assistance data transfer, or location information transfer.
LibSUPL supports only assistance data transfer as the transaction type.

The following image shows the steps in the libSUPL connection session.

> .. figure:: /images/libSUPL_connection_session.png

Reviewer 2: We should replace this image with an MSC if possible - I'll try to come up with the code. :)

Reviewer 2: Try this here - not perfect, but I think it'll do:
```

.. msc::
   hscale = "1.3";
   SET,SLP;
   --- [label="Data connection setup"];
   SET=>SLP      [label="SUPL START (session-id, lid, SET capabilities"];
   SET<=SLP      [label="SUPL RESPONSE (session-id, posmethod)"];
   SET=>SLP      [label="SUPL POS INIT (session-id, lid, SET capabilities"];
   SET..SLP [linecolor="#00a9ce", textcolor="#00a9ce", label="\nloop 1..N times"];
   SET<=SLP      [label="SUPL POS (session-id, LPP)"];
   ...;
   SET..SLP [linecolor="#00a9ce"];
   SET<<SLP      [label="SUPL END (session-id)"];
   --- [label="Data connection teardown"];

```

Author: Replaced the image with MSC. Thanks Ruth!

      :alt: LibSUPL connection session

> LibSUPL architecture

Reviewer 2: ```suggestion
SUPL client architecture
```

(That's why I asked for the other sections if they are specific to this "SUPL client" or if they are generic - this here is clearly specific, the other ones seem more general to me.)

********************

The libSUPL architecture consists of modules (both internal and external) and API callback functions.
These are described in the subsequent sections.

Modules
=======

There are several modules inbuilt into the LibSUPL architecture as shown in the figure below.
Additionally, the library makes use of the external module :ref:`ossnokalva_asn1_lib`.

> .. figure:: /images/libsupl_module_architecture.png

Author: Replacing with a new image.

      :alt: LibSUPL module architecture

Resource initialization and ownership
+++++++++++++++++++++++++++++++++++++

The buffers used in the SUPL communication session, namely the primary and secondary buffers, are owned by the user of the libSUPL library.
The buffers are characterized by the sizes defined by LIBSUPL_PRI_BUF_SIZE and LIBSUPL_SEC_BUF_SIZE.
These parameters are defined in the ``supl_session.h`` header file.

> The buffers are passed to libSUPL through the supl_client_init() function.

Reviewer: As a NCS user you should preferably use the supl_init() function instead of the supl_client_init() function. On the other hand if you are a non NCS user you will have to use the supl_client_init() function. I don't know how and if we should distinguish between these use-cases in this document. supl_init is the API for the integration part, while supl_client_init is the library init function.

> The buffers are passed to libSUPL through the supl_client_init() function.

Reviewer 2: We don't need to document anything for non-NCS users.
So it should probably be supl_init() instead of supl_client_init() here?

Do NCS users need to know about any functions in the external SUPL library?

Communication with the SUPL server must be handled by the application through the read and write callback functions.

SUPL Flow Control
+++++++++++++++++

> LibSUPL is initialized by calling the supl_init() function before the first SUPL session.

Reviewer 2: :cpp:func:`supl_init` *should* work (same for supl_session).

Author: Made this change for all functions specified.

The supl_session() function forms the entry point for the SUPL session.
There is no specific limit on the number of subsequent SUPL sessions that can be executed from this entry point.
See the documentation on the `SET-Initiated Non-Roaming Successful Case (Proxy mode)`_ for more information on the SUPL session.

The figure below shows the flow of communication in the SUPL session.

> .. figure:: /images/supl_session.png

Author: Replacing with a new image.

      :alt: SUPL session

The user of the libSUPL library can request premature exit from the SUPL session by calling the supl_abort() function.
In practice, the abort flag is checked on every return from the SUPL Read callback function.

.. _ossnokalva_asn1_lib:

OSS Nokalva ASN.1 (Abstract Syntax Notation One) library
++++++++++++++++++++++++++++++++++++++++++++++++++++++++

`OSS Nokalva ASN.1 library`_ is a third-party library, which is used by libSUPL for decoding and encoding SUPL and LPP messages.
LibSUPL uses the time-optimized version of the library as it produces the smallest code footprint with libSUPL related message definitions.

Buffer Management
+++++++++++++++++

In this library, two separate buffers (the primary buffer and the secondary buffer) are needed to store encoded and decoded formats of the data concurrently.
The following sequential diagram and steps show the receiving and sending of the SUPL messages.

> .. figure:: /images/supl_buffer_seqdgm.png

Author: Replacing with a new image.

      :alt: SUPL session

Receiving

>    The SUPL Read callback function receives the encoded SUPL message.

Reviewer: Should we mention that this all of the callbacks must be implemented by the user. And it's done like this to let the library make use of the platform specific communication APIs.

Sending

>    The SUPL write callback function sends the encoded SUPL message.

Reviewer 2: ```suggestion
   The SUPL Write callback function sends the encoded SUPL message.
```

   As of now, libSUPL library does not include the LPP payload in any of the messages sent.
   Hence it utilizes only the primary buffer during encoding and sending of the encoded SUPL data.

> 

Reviewer 2: Not sure I understand where the decoding happens? In the callback function, and the callback function can use the different buffers?

Reviewer: The sending callback doesn't know anything about the buffers, it is only given a pointer to a buffer and a length. I don't think we need mention the buffers here. In NCS the buffers are take care of by the integration library, so it's completely transparent for the user. If we should be mention them it should be in a section about initialization.

Author: @evenl I assume that we are removing the entire 'Buffer management' section here..

LPP2NRF
+++++++

> This module converts SUPL structures into nRF Assisted GNSS structures.

Reviewer 2: What are nRF Assisted GNSS structures?

Reviewer: Yes, sounds a bit strange. maybe something like " ... into data structures defined nrf_socket.h used by the nrf GNSS socket?

Reviewer 2: Are they documented somewhere here? https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/bsdlib/doc/api.html#gnss-socket
Then we could link.

Pointers of the nRF AGNSS structures are forwarded to the AGPS Handler callback function.
Pointers are valid till the return of the callback function, after which they fall out of scope.

> API Callback Functions

Reviewer 2: ```suggestion
API callback functions
```

======================

Callback functions are needed to achieve platform independence.

> Callback functions are implemented by the user of the libSUPL library.

Reviewer 2: Are there any example implementations? Probably in the GPS sample, or somewhere more general?

Reviewer: Yes, the GPS sample has support for SUPL. It has a separate file that implements these callback functions.

The various callback functions implemented in this library and their purposes are listed below.

AGPS Handler
   Callback function to handle nRF assistance data types.

SUPL logger
   Callback function to output libSUPL debug messages.
   Callback implementation is optional.
   If the callback function is not implemented, the library does not display debug messages.

Millisecond Counter
   Callback function for millisecond counter.
   Callback implementation is optional.

>    If the callback function is not implemented, the library proceeds with 1 second socket receive timeout and estimates SUPL level timeouts based on that.

Reviewer 2: ```suggestion
   If the callback function is not implemented, the library proceeds with 1 second socket receive time-out and estimates SUPL level time-outs based on that.
```
